### PR TITLE
Add sha512 generation to promotion job

### PR DESCRIPTION
**Pull request by peternied**
### Description
The promotion workflow is picking up artifacts from a distribution build and placing them in their final resting place on the artifacts server.  This is a perfect opportunity to do any final polishing steps on artifacts that are ready for release.  Adding this hook for file actions and creating a sha512 generation closure.

This is untested on jenkins, but is the core POC needed to bootstrap the signing workflow.
 
### Issues Resolved
- https://github.com/opensearch-project/opensearch-build/issues/1351
- https://github.com/opensearch-project/opensearch-build/issues/1355
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: vars/createSha512.groovy [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Closure call() {
```
Comment (peternied) [inline vars/createSha512.groovy:1]: This was a quick'n dirty method to pass around these functions calls - there might be a better way

File: tests/jenkins/TestPromoteArtifacts.groovy [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Test
+    public void testDefault() {
+        super.testPipeline("tests/jenkins/jobs/PromoteArtifacts_Jenkinsfile")
+    }
+
+    @Test
+    public void testWithActions() {
+        super.testPipeline("tests/jenkins/jobs/PromoteArtifacts_actions_Jenkinsfile")
+    }
```
Comment (peterzhuamazon) [inline tests/jenkins/TestPromoteArtifacts.groovy:71]: Nice, so we can have with and without actions testing in both.

File: tests/jenkins/jobs/PromoteArtifacts_actions_Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
+pipeline {
+    agent none
+    stages {
+        stage('promote') {
+            steps {
+                script {
+                    promoteArtifacts(
+                        manifest: 'tests/jenkins/data/opensearch-1.3.0.yml',
+                        fileActions: [createSha512()]
```
Comment (peterzhuamazon) [inline tests/jenkins/jobs/PromoteArtifacts_actions_Jenkinsfile:9]: Assuming we can expand this into `fileActions: [createSha512(), createPgpSig()]`

Comment (peterzhuamazon) [reply tests/jenkins/jobs/PromoteArtifacts_actions_Jenkinsfile:9]: #1352

File: vars/createSha512.groovy (hunk 1 above)
Comment (peterzhuamazon) [reply vars/createSha512.groovy:1]: ELI5: what is the reason that we cannot use a normal `call` here?
From what I read it is meant for simplistic and able to passed as argument?
Therefore you can pass `createSha512` as an argument in promoteArtifacts?

File: vars/promoteArtifacts.groovy [hunk 4]
```diff
@@ -32,22 +37,47 @@ void call(Map args = [:]) {
             String pluginSubFolder = pluginSubPath.split('/')[0]
             String pluginNameWithExt = pluginSubPath.split('/')[1]
             String pluginName = pluginNameWithExt.replace('-' + version + '.zip', '')
+            String pluginNameNoExt = pluginNameWithExt.replace('-' + version, '')
             String pluginFullPath = ['plugins', pluginName, version].join('/')
-            s3Upload(bucket: "${ARTIFACT_PRODUCTION_BUCKET_NAME}", path: "releases/$pluginFullPath/", workingDir: "$WORKSPACE/artifacts/$artifactPath/builds/$filename/$pluginSubFolder/"
-                , includePathPattern: "**/${pluginName}*")
+            for (Closure action : fileActions) {
+                for (file in findFiles(glob: "$WORKSPACE/artifacts/$artifactPath/builds/$filename/$pluginSubFolder/**/${pluginName}*")) {
```
Comment (peterzhuamazon) [inline vars/promoteArtifacts.groovy:43]: We can merge here but as #1355 points out we need to save these params into their own var.

**Review by peterzhuamazon — APPROVED** [review]
Overall great addition to the promotion workflow and allow us to expand with signer and more actions later.

Thanks!

File: vars/promoteArtifacts.groovy (hunk 4 above)
Comment (peternied) [reply vars/promoteArtifacts.groovy:43]: This PR doesn't resolve #1355, but it did some of the things called out there so I figured it was better to mention it.

File: vars/createSha512.groovy (hunk 1 above)
Comment (peternied) [reply vars/createSha512.groovy:1]: Not quite the ELI5 level, this is just copying prior art and convention.  `call(...)` is the method that is invoked when you do `createSha512(...)`.  Its signature supports parameters, typically we accept a map of values and stuff it into an object called args.  Typically we don't return anything, making the return type void.  So this function is following the same generalized pattern, but it doesn't accept parameters (we could add that) and it returns a closure that can be used.

**Review by dblock — APPROVED** [review]
I like it! You definitely want to test in Jenkins because permissions might not let you do what you're trying to do :(

**Review by dblock — APPROVED** [review]
